crates/dekaf/src/log_journal.rs

```diff
@@ -0,0 +1,344 @@
+// messages, both buffered and ongoing, will get written to the logs journal
+// associated with that task.
+pub const SESSION_SPAN_NAME_MARKER: &str = "dekaf_session";
+// When a span is created with this field name, and is a descendent of a span
+// named with the value of `SESSION_SPAN_NAME_MARKER`, it causes events (logs)
+// anywhere in that hierarchy to get written to the corresponding logs journal.
+pub const SESSION_TASK_NAME_FIELD_MARKER: &str = "session_task_name";
+// This marker indicates that its Session is not and will never be associated with a
+// task, and we should stop buffering logs as we'll never have anywhere to write them.
+pub const SESSION_TASKLESS_FIELD_MARKER: &str = "session_is_taskless";
+
+#[derive(Debug)]
+enum LoggingMessage {
+    SetTaskName(String),
+    Log(ops::Log),
+    Shutdown,
+}
+
+struct LogForwarder {
+    app: Arc<App>,
+    producer: gazette::uuid::Producer,
+}
+
+impl LogForwarder {
+    fn new(app: Arc<App>, producer: Producer) -> Self {
+        Self { app, producer }
+    }
+
+    async fn forward_logs(
+        self,
+        mut logs_rx: tokio::sync::mpsc::Receiver<LoggingMessage>,
+    ) -> anyhow::Result<()> {
+        let mut log_data = Vec::new();
+
+        let (ops_logs_journal_client, ops_logs_journal) = loop {
+            match logs_rx.recv().await {
+                Some(LoggingMessage::SetTaskName(name)) => {
+                    let (client, ops_logs) = self.get_journal_client(name).await?;
+                    break (client, ops_logs);
+                }
+                Some(LoggingMessage::Log(log)) => {
+                    log_data.append(&mut self.serialize_log(log));
+                }
+                Some(LoggingMessage::Shutdown) | None => return Ok(()),
+            }
+        };
+
+        self.append_log_data(
+            log_data.into(),
+            ops_logs_journal.as_str(),
+            &ops_logs_journal_client,
+        )
+        .await?;
+
+        while let Some(msg) = logs_rx.recv().await {
+            match msg {
+                LoggingMessage::SetTaskName(_) => {}
+                LoggingMessage::Log(log) => {
+                    self.append_log_data(
+                        self.serialize_log(log).into(),
+                        ops_logs_journal.as_str(),
+                        &ops_logs_journal_client,
+                    )
+                    .await?;
+                }
+                LoggingMessage::Shutdown => break,
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn append_log_data(
+        &self,
+        log_data: Bytes,
+        journal: &str,
+        client: &journal::Client,
+    ) -> anyhow::Result<()> {
+        let resp = client.append(
+            gazette::broker::AppendRequest {
+                journal: journal.to_owned(),
+                ..Default::default()
+            },
+            || {
+                futures::stream::once({
+                    let value = log_data.clone();
```

Comment:
Pretty sure this is a cheap clone since `log_data` is `Bytes` and never modified. It needs to be cloned because this is `Fn`, not `FnOnce` since it needs to be callable multiple times for retry purposes